### DISCUSSION TITLE
Unhandled errors thrown during execution of a prepared statement

### DISCUSSION THREAD

**Discussion by jeroenrinzema** [jackc/pgx#1466] [Ideas]
👋🏻 , I am currently working on a PSQL server wire protocol implementation allowing you to create a [PSQL server](https://github.com/jeroenrinzema/psql-wire) in a few lines of code. I noticed while testing the extended query protocol that errors thrown while executing a prepared statement are not handled. I have opened a [PR](https://github.com/jeroenrinzema/psql-wire/pull/53) inside the `psql-wire` protocol repository which is able to reproduce the issue.

It seems that the error is not handled due to the error not being handled when reading until a row descriptor while executing a prepared query.

https://github.com/jackc/pgx/blob/74f9b9f0a483f95513c621364f2c3912181ee360/pgconn/pgconn.go#L1480-L1484

I would like to propose to handle the error returned when reading the next message and to return it to the client.

**Comment by jackc** [comment]
I don't understand what's going on here. The call to `rr.receiveMessage` should handle the `ErrorResponse`.

https://github.com/jackc/pgx/blob/fcec008a4cc906b9261a66f09e2248a128ff2282/pgconn/pgconn.go#L1516-L1525

That will terminate the loop in `readUntilRowDescription`.

AFAICT pgx is doing the right thing. Can you compare the message traffic with psql-wire to PostgreSQL? (pgproto3.Frontend.Trace might be handy here)

**Comment by jeroenrinzema** [comment]
Thanks for looking into this! I have been diving a bit deeper into the message flow. The UNIT test I have written returned a `FATAL` error message causing an error to be returned while receiving a message from the [pg conn](https://github.com/jackc/pgx/blob/fcec008a4cc906b9261a66f09e2248a128ff2282/pgconn/pgconn.go#L502-L508). The interesting thing is that errors returned during evaluation are not handled within the returned rows. The error is set correctly within the [result reader](https://github.com/jackc/pgx/blob/fcec008a4cc906b9261a66f09e2248a128ff2282/conn.go#L758) but is not handled inside the [`baseRows`](https://github.com/jackc/pgx/blob/fcec008a4cc906b9261a66f09e2248a128ff2282/rows.go#L187-L189) type. I have tested this with errors of different severity levels.

I have compared the message flow of psql-wire to PostgreSQL but was unable to reproduce the throwing of an error during the execution of a prepared statement.  Please let me know if you have suggestions on how I could replicate this error message flow inside PostgreSQL.

**Comment by jackc** [comment]
I just looked a little closer at your test. The problem is the query results are not being read. 

This test works:

```go
	t.Run("jackc/pgx", func(t *testing.T) {
		ctx := context.Background()
		connstr := fmt.Sprintf("postgres://%s:%d", address.IP, address.Port)
		conn, err := pgx.Connect(ctx, connstr)
		assert.NoError(t, err)

		rows, _ := conn.Query(ctx, "SELECT *;")
		rows.Close()
		assert.Error(t, rows.Err())

		err = conn.Close(ctx)
		assert.NoError(t, err)
	})
```

**Comment by jeroenrinzema** [comment]
Perfect! Thanks for looking into my tests. I have updated the test and everything seems to be working perfectly! We could convert this discussion to an issue and close it if you like.
